### ISSUE TITLE
GPT always overwritten by zpool create of a whole disk with a mounted ext4 partition 1

### ISSUE THREAD

**Issue by ilovezfs** [zfsonlinux/zfs#4016]
ZoL's zpool create always overwrites GPT of a whole disk with a mounted ext4 file system on partition 1 even with liblkid and without -f.

Situation: The first partition of a disk has a mounted ext4 file system.
openzfs ~ # mount -t ext4 /dev/sdc1 /mnt/BusyExt4/
openzfs ~ # mount | grep sdc
/dev/sdc1 on /mnt/BusyExt4 type ext4 (rw,relatime,data=ordered)

Problem commands:
zpool create thiswontwork sdc
zpool create thiswontwork /dev/sdc
zpool create -f thiswontwork sdc
zpool create -f thiswontwork /dev/sdc

Result: the GPT is overwritten with the standard ZFS partitioning scheme. The ext4 file system is recoverable if a partition table using the same offsets as the original is restored.

This is a regression caused by "Fix hot spares" https://github.com/zfsonlinux/zfs/commit/8128bd89fb487d91a3335605b009ac2b7d5aecd3#diff-ea3469b2eb70e3b5a05c22cf1a8ed06bL356

It is still present in current master https://github.com/zfsonlinux/zfs/blob/master/cmd/zpool/zpool_vdev.c#L541

The problem is that ZFS expects construct_spec() to have noticed the EBUSY condition:
https://github.com/zfsonlinux/zfs/blob/master/cmd/zpool/zpool_vdev.c#L1696-L1702

Upstream construct_spec() relies on the behavior of the illumos is_whole_disk()'s use of O_RDWR | O_NDELAY in order to detect the busy condition:
https://github.com/illumos/illumos-gate/blob/master/usr/src/cmd/zpool/zpool_vdev.c#L353-L371

Note that there is a secondary less serious bug with these commands in the same situation:
zpool create thiswontwork -f sdc1
zpool create thiswontwork -f /dev/sdc1

We get an incorrect error message: "cannot create 'thiswontwork': one or more vdevs refer to the same device, or one of the devices is part of an active md or lvm device" because with -f it doesn't take advantage of liblkid for some reason. I've not looked into where that goes awry.

Demonstration of both bugs:

```
openzfs ~ # modinfo spl zfs | grep '0.6.'
version:        0.6.5-7_g3e7e6f3
version:        0.6.5-31_gf3e2a7a

//ZoL, whole device, short name, no force:
openzfs ~ # zpool create thiswontwork sdc
the kernel failed to rescan the partition table: 16
cannot label 'sdc': try using parted(8) and then provide a specific slice: -1

GPT: Partition table overwritten despite no -f.
openzfs ~ # lsblk /dev/sdc
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sdc      8:32   0    1G  0 disk 
└─sdc1   8:33   0 1023M  0 part /mnt/BusyExt4
openzfs ~ # partprobe
openzfs ~ # lsblk /dev/sdc
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sdc      8:32   0    1G  0 disk 
├─sdc1   8:33   0 1014M  0 part /mnt/BusyExt4
└─sdc9   8:41   0    8M  0 part 

Second try:
openzfs ~ # zpool create thiswontwork sdc
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 contains a filesystem of type 'ext4'
(Also, note that it doesn't notice that partitition 1 on disk was what it wrote, and is at different offsets than the mounted ext4. We get a different error the second time because ZFS sees the standard ZFS partitioning on disk already.)

//ZoL, whole device, long name, no force:
openzfs ~ # zpool create thiswontwork /dev/sdc
      Same as ZoL, whole device, short name, no force

//ZoL, whole device, short name, force:
openzfs ~ # zpool create -f thiswontwork sdc
the kernel failed to rescan the partition table: 16
cannot label 'sdc': try using parted(8) and then provide a specific slice: -1
(Note the error is the same as last time. -f makes no difference)
GPT: same as without force

Second try: Same error as first try (unlike in the no force case)
openzfs ~ # partprobe
openzfs ~ # zpool create -f thiswontwork sdc
the kernel failed to rescan the partition table: 16
cannot label 'sdc': try using parted(8) and then provide a specific slice: -1

But if we now try again without force, we'll get the same error as we got on the second try in the no force case last time.
openzfs ~ # zpool create thiswontwork sdc
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 contains a filesystem of type 'ext4'

//ZoL, whole device, long name, force:
openzfs ~ # zpool create -f thiswontwork /dev/sdc
    Same as ZoL, whole device, short name, force

//ZoL, partition, short name, no force:
openzfs ~ # zpool create thiswontwork sdc1
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 contains a filesystem of type 'ext4'

GPT: unchanged
2nd try: no difference

//ZoL, partition, long name, no force:
    Same as ZoL, partition, short name, no force

//ZoL, partition, short name, force:
openzfs ~ # zpool create thiswontwork -f sdc1
cannot open '/dev/sdc1': Device or resource busy
cannot create 'thiswontwork': one or more vdevs refer to the same device, or one of
the devices is part of an active md or lvm device

GPT: No impact
2nd try: No difference

//ZoL, partition, long name, force:
openzfs ~ # zpool create thiswontwork -f /dev/sdc1
    Same as ZoL, partition, short name, force
```
 (Note: I stumbled upon this bug while investigating https://github.com/openzfsonosx/zfs/issues/415)

**Comment by ilovezfs** [comment]
Additional research (https://gist.github.com/ilovezfs/81543139627a1b084dde) shows illumos doesn't expect is_whole_disk() to return false in this situation. Prior to https://github.com/zfsonlinux/zfs@8128bd8#diff-ea3469b2eb70e3b5a05c22cf1a8ed06bL356 and its switch from O_RDWR|O_DIRECT|O_EXC to O_RDONLY|O_DIRECT, ZoL would (incorrectly) return false there, which would then (accidentally) have the salutary effect of preventing the GPT from being overwritten because it would go down the same code path as the partitions, which as shown above, don't lead to an overwriting of the GPT (no need to auto-partition a partition).

The busy situation should be caught before we ever get near calling efi_write(), but putting that aspect of the bug aside for the moment, the immediate proximate cause of the GPT overwriting is here:
https://github.com/zfsonlinux/zfs/blob/b815ec32b3ba04ad7307ca8945a865eec0ec766b/lib/libzfs/libzfs_pool.c#L4169
<code>	if ((fd = open(path, O_RDWR|O_DIRECT)) < 0) {</code>

as we can see by running
```
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <errno.h>

void
try_open(const char *path, int oflag)
{
	int fd;
	errno = 0;
	if ((fd = open(path, oflag)) < 0) {
		printf("fail : errno %d\n", errno);
	} else {
		printf("pass\n");
		(void) close(fd);
	}
}

int
main(int argc, char *argv[])
{
	if (argc < 2) {
		printf("not enough args\n");
		return (2);
	}

	const char *path = argv[1];

	printf("Trying to open %s O_RDONLY|O_DIRECT ...\n", path);
	try_open(path, O_RDONLY|O_DIRECT);

	printf("Trying to open %s O_RDWR|O_DIRECT ...\n", path);
	try_open(path, O_RDWR|O_DIRECT);

	printf("Trying to open %s O_RDWR|O_DIRECT|O_EXCL ...\n", path);
	try_open(path, O_RDWR|O_DIRECT|O_EXCL);

	return (0);
}
```
which gives us
```
openzfs ~ # ./busy /dev/sdc
Trying to open /dev/sdc O_RDONLY|O_DIRECT ...
pass
Trying to open /dev/sdc O_RDWR|O_DIRECT ...
pass
Trying to open /dev/sdc O_RDWR|O_DIRECT|O_EXCL ...
fail : errno 16
openzfs ~ # ./busy /dev/sdc1
Trying to open /dev/sdc1 O_RDONLY|O_DIRECT ...
pass
Trying to open /dev/sdc1 O_RDWR|O_DIRECT ...
pass
Trying to open /dev/sdc1 O_RDWR|O_DIRECT|O_EXCL ...
fail : errno 16
```

So one stop gap would be changing the O_RDWR|O_DIRECT to O_RDWR|O_DIRECT|O_EXCL in zpool_label_disk() so that the open fails. Currently, the open succeeds and efi_write() is called with the fd.

There is likely an equivalent bug here:
https://github.com/zfsonlinux/zfs/blob/b815ec32b3ba04ad7307ca8945a865eec0ec766b/lib/libzfs/libzfs_pool.c#L2380
<code>	if ((fd = open(path, O_RDWR|O_DIRECT)) < 0) {</code>
because that fd is passed to efi_use_whole_disk() which then calls efi_write() with the same fd:
https://github.com/zfsonlinux/zfs/blob/master/lib/libefi/rdwr_efi.c#L1106

Perhaps it would be safer to have efi_write() itself check to make sure the fd is opened exclusively, or make it open its own fd exclusively after closing the one it is handed, since efi_write() can be called directly and indirectly (via efi_use_whole_disk()) from libzfs_pool.c and the caller is apparently responsible for making sure the fd is opened exclusively, which isn't obvious.

**Comment by ilovezfs** [comment]
@behlendorf Mystery solved.

The problem is specifically partition type 8300 Linux filesystem (0FC63DAF-8483-4772-8E79-3D69D8477DE4), as well as any other partition type not in the conversion array https://github.com/zfsonlinux/zfs/blob/5c790678f1a5231b75be174eb8a02e15dfe9b4d5/lib/libefi/rdwr_efi.c#L48-L73

This diff resolves the bug for 8300 specifically:
```
diff --git a/include/sys/efi_partition.h b/include/sys/efi_partition.h
index ee367a5..1cf7a21 100644
--- a/include/sys/efi_partition.h
+++ b/include/sys/efi_partition.h
@@ -134,6 +134,8 @@ typedef struct efi_gpe_Attrs {
 			    { 0x00, 0x30, 0x65, 0x43, 0xec, 0xac } }
 #define	EFI_AAPL_UFS	{ 0x55465300, 0x0000, 0x11aa, 0xaa, 0x11, \
 			    { 0x00, 0x30, 0x65, 0x43, 0xec, 0xac } }
+#define EFI_DELL_LINUX	{ 0x0fc63daf, 0x8483, 0x4772, 0x8e, 0x79, \
+			    { 0x3d, 0x69, 0xd8, 0x47, 0x7d, 0xe4 } }
 
 /* minimum # of bytes for partition table entires, per EFI spec */
 #define	EFI_MIN_ARRAY_SIZE	(16 * 1024)
diff --git a/lib/libefi/rdwr_efi.c b/lib/libefi/rdwr_efi.c
index 785cfe6..a5c0044 100644
--- a/lib/libefi/rdwr_efi.c
+++ b/lib/libefi/rdwr_efi.c
@@ -68,6 +68,7 @@ static struct uuid_to_ptag {
 	{ EFI_DELL_SWAP },
 	{ EFI_DELL_LVM },
 	{ EFI_DELL_RESV },
+	{ EFI_DELL_LINUX},
 	{ EFI_AAPL_HFS },
 	{ EFI_AAPL_UFS }
 };
```
The problem occurs here:
https://github.com/zfsonlinux/zfs/blob/5c790678f1a5231b75be174eb8a02e15dfe9b4d5/cmd/zpool/zpool_vdev.c#L474-L476

Since there is no match in the table, vtoc->efi_parts[i].p_tag == V_UNASSIGNED for all 128 iterations, and we end up exhausting the loop without ever calling check_slice().

@behlendorf other than adding as many known partition types as possible (https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_type_GUIDs), any thoughts on how you would like this bug addressed?

**Comment by behlendorf** [comment]
@ilovezfs nice analysis!  Well I think we should definitely start by adding at least the most common additional partition types.  That's straight forward and will probably fix the most common instances of this.    I'm open to suggestions about how to fix this more generically. Do you have anything specific in mind?

**Comment by ilovezfs** [comment]
@behlendorf I've added a panoply of new types here:
Added additional types https://github.com/zfsonlinux/zfs/pull/4032

I assume many/most deserve to be vetoed but did not try to narrow the list down to an optimal set myself. Hopefully, the names are consistent with the prior naming convention.

In terms of the underlying issue, I was thinking maybe ZoL should rely more heavily on the mount table, blkid, etc. before worrying about the partition type on the efi label. illumos does all of these checks when it calls dm_inuse on a slice:

```
static detectorp detectors[] = {
	inuse_mnt,
	inuse_svm,
	inuse_active_zpool,
	inuse_lu,
	inuse_dump,
	inuse_vxvm,
	inuse_exported_zpool,
	inuse_fs,  /* fs should always be last */
	NULL
};
```

https://github.com/illumos/illumos-gate/blob/139510fb6efa97dbe5f5479594b308d940cab8d1/usr/src/lib/libdiskmgt/common/slice.c#L57-L67

https://github.com/illumos/illumos-gate/blob/28b6fd27d5ff75fe6fdeb119a21575b0652a7e70/usr/src/cmd/zpool/zpool_vdev.c#L146

**Comment by behlendorf** [comment]
@ilovezfs I don't think we need to veto any of them.  They're all seem to be valid, just not common, but that't not really a problem.  I've merged #4032 to master.

I completely agree we should be relying more heavily on blkid.  We should probably make using it mandatory and not optional.  It would be great if you could work on improving and testing the existing HAVE_LIBBLKID support.

**Comment by behlendorf** [comment]
Closing.  Additional partition types have been added and blkid is now mandatory.

7d11e37 Require libblkid
917b8c5 Ext4's typical GPT partition type not recognized
fc80384 Illumos 934 - FreeBSD's GPT not recognized
